ludwig/benchmarking/profiler.py

```diff
@@ -0,0 +1,271 @@
+            for i, gpu_info in enumerate(gpu_infos):
+                gpu_key = f"cuda_{i}"
+                info[f"{gpu_key}_memory_used"].append(gpu_info.memory_used)
+        with tracked_process.oneshot():
+            info["cpu_utilization"].append(tracked_process.cpu_percent())
+            info["cpu_memory_usage"].append(tracked_process.memory_full_info().uss)
+        time.sleep(logging_interval)
+
+
+class LudwigProfiler(contextlib.ContextDecorator):
+    """Track system resource (hardware and software) usage.
+
+    Warning: If `use_torch_profiler=True` while profiling on CUDA, it's not possible to benchmark DataLoaders
+    with `num_workers > 0` due to CUDA multiprocessing limitations. See warning under `profile` class
+    definition: https://github.com/pytorch/pytorch/blob/master/torch/autograd/profiler.py
+
+    Attributes:
+        tag: a string tag describing the code block/function that we're tracking.
+            (e.g trainer.train, preprocessing, etc.)
+        output_dir: path where metrics are saved.
+        logging_interval: time interval in seconds at which system is polled for resource usage.
+    """
+
+    def __init__(self, tag: str, use_torch_profiler: bool, output_dir: str, logging_interval: float = 0.1) -> None:
+        self.tag = tag
+        self._tag = LUDWIG_TAG + self.tag
+        self.use_torch_profiler = use_torch_profiler
+        self.output_dir = output_dir
+        self.logging_interval = logging_interval
+        self.cuda_is_available = torch.cuda.is_available()
+        self.launched = False
+        if self.use_torch_profiler:
+            self.profiler_activities = [torch.profiler.ProfilerActivity.CPU]
+            if self.cuda_is_available:
+                self.profiler_activities.append(torch.profiler.ProfilerActivity.CUDA)
+        os.makedirs(os.path.join(self.output_dir), exist_ok=True)
+
+    def _init_tracker_info(self):
+        """Initialize new self.info, self.torch_profiler, and self.torch_record_function instances.
+
+        Important to call this in __enter__ if the user decides not to create a new class instance and therefore
+        __init__ wouldn't be called.
+        """
+        self.info = {"code_block_tag": self.tag}
+        if self.use_torch_profiler:
+            self.torch_profiler = torch.profiler.profile(activities=self.profiler_activities, profile_memory=True)
+            self.torch_record_function = torch.profiler.record_function(self._tag)
+
+    def populate_static_information(self) -> None:
+        """Populate the report with static software and hardware information."""
+        self.info["ludwig_version"] = LUDWIG_VERSION
+        self.info["start_disk_usage"] = shutil.disk_usage(os.path.expanduser("~")).used
+
+        # CPU information
+        cpu_info = get_my_cpu_info()
+        self.info["cpu_architecture"] = cpu_info["arch"]
+        self.info["num_cpu"] = cpu_info["count"]
+        self.info["cpu_name"] = cpu_info["brand_raw"]
```

Comment:
Fails on local when running tests with:
```
>       self.info["cpu_name"] = cpu_info["brand_raw"]
E       KeyError: 'brand_raw'
```

Reply:
Wasn't able to reproduce this. Let me know if you run into this again.